File: Game_of_life/js/Game_of_life.js
const generationsToRender = 10;
const matrix = [
  [0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 1, 0, 0, 0, 0, 0, 0, 0],
  [1, 1, 1, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
];

function printMatrix(matrixToPrint) {
  for (var i = 0; i < matrixToPrint.length; i++) {
    console.log(matrixToPrint[i].join(","));
  }
}

function computeNextGeneration(matrixToCompute) {
  var nextGenerationMatrix = [];
  for (var i = 0; i < matrixToCompute.length; i++) {
    var row = matrixToCompute[i];
    var newRow = [];
    for (var j = 0; j < row.length; j++) {
      newRow.push(computeNextGenerationStatus(matrixToCompute, i, j));
    }
    nextGenerationMatrix.push(newRow);
  }
  return nextGenerationMatrix;
}

function computeNextGenerationStatus(matrixToCompute, x, y) {
  var neighbours = findNeighbours(matrixToCompute, x, y);
  if (neighbours == 0 || neighbours == 1) {
    return 0;
  }

  if (neighbours == 2) {
    return matrixToCompute[x][y];
  }

  if (neighbours == 3) {
    return 1;
  }

  if (neighbours > 3) {
    return 0;
  }
}

function findNeighbours(matrixToCompute, x, y) {
  var neighbourCount = 0;
  for (var i = x - 1; i <= x + 1; i++) {
    if (i == -1 || i == matrixToCompute.length) {
      continue;
    }

    for (var j = y - 1; j <= y + 1; j++) {
      if (j == -1 || j == matrixToCompute.length) {
        continue;
      }

      if (i == x && j == y) {
        continue;
      }

      if (matrixToCompute[i][j] != undefined) {
        matrixToCompute[i][j] == 1 ? neighbourCount++ : "";
      }
    }
  }

  return neighbourCount;
}

(function() {
  console.log("Initial state.");
  console.log();
  printMatrix(matrix);
  var m = matrix;
  for (var i = 1; i <= generationsToRender; i++) {
    console.log("Generation " + i);
    m = computeNextGeneration(m);
    printMatrix(m);
    console.log();
  }
})();

File: interesting_algorithms/Game_of_life/dart/Game_of_life.dart
main() {
  int generationsToRender = 10;
  List<List<int>> matrix = [
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
  ];

  renderGeneration(matrix, generationsToRender);
}

void renderGeneration(List<List<int>> matrix, int generatonsToRender) {
  print("Initial state");
  printMartix(matrix);
  print("");

  for (int i = 1; i <= generatonsToRender; i++) {
    matrix = generateNextGenerations(matrix);
    print("Generation ${i}");
    printMartix(matrix);
    print("");
  }
}

void printMartix(List<List<int>> matrixToPrint) {
  matrixToPrint.forEach((List<int> row) => {print(row.join(","))});
}

List<List<int>> generateNextGenerations(List<List<int>> matrix) {
  List<List<int>> nextGenerationMatrix = [];

  for (int x = 0; x < matrix.length; x++) {
    List<int> row = matrix[x];
    List<int> newRow = [];

    for (int y = 0; y < row.length; y++) {
      newRow.add(computeNextGenerationStatus(matrix, x, y));
    }
    nextGenerationMatrix.add(newRow);
  }

  return nextGenerationMatrix;
}

int computeNextGenerationStatus(List<List<int>> matrix, int x, int y) {
  int neighboursCount = findNeighbours(matrix, x, y);

  if (neighboursCount <= 1) {
    return 0;
  }

  if (neighboursCount == 2) {
    return matrix[x][y];
  }

  if (neighboursCount == 3) {
    return 1;
  }

  return 0;
}

int findNeighbours(List<List<int>> matrix, int x, int y) {
  int neighboursCount = 0;
  for (int _x = x - 1; _x <= x + 1; _x++) {
    if (_x == -1 || _x == matrix.length) {
      continue;
    }

    for (int _y = y - 1; _y <= y + 1; _y++) {
      if (_y == -1 || _y == matrix.length || (_x == x && _y == y)) {
        continue;
      }

      matrix[_x][_y] == 1 ? neighboursCount++ : "";
    }
  }

  return neighboursCount;
}
